sklearn/multioutput.py

```diff
@@ -648,6 +648,51 @@ def _log_message(self, *, estimator_idx, n_estimators, processing_msg):
             return None
         return f"({estimator_idx} of {n_estimators}) {processing_msg}"
 
+    def _get_chain_method(self):
+        # Different chain methods allowed in `ClassifierChain` only
+        try:
+            chain_method = self.chain_method_
+        except AttributeError:
+            chain_method = "predict"
+        return chain_method
+
+    def _get_Y_output(self, X, *, output_method="predict"):
+        """Get predictions for each model in the chain."""
+        check_is_fitted(self)
+        X = self._validate_data(X, accept_sparse=True, reset=False)
+        Y_output_chain = np.zeros((X.shape[0], len(self.estimators_)))
+        Y_feature_chain = np.zeros((X.shape[0], len(self.estimators_)))
+
+        chain_method = self._get_chain_method()
+        # proba methods produce 2d output (decision_function 1d for binary targets)
+        is_feature_multi = chain_method in ["predict_proba", "predict_log_proba"]
+        is_output_multi = output_method == "predict_proba"
+
+        for chain_idx, estimator in enumerate(self.estimators_):
+            previous_predictions = Y_feature_chain[:, :chain_idx]
+            if sp.issparse(X):
+                X_aug = sp.hstack((X, previous_predictions))
+            else:
+                X_aug = np.hstack((X, previous_predictions))
```

Comment:
```suggestion
            hstack = sp.hstack if sp.issparse(X) else np.hstack
            X_aug = hstack([X, previous_prediction])
```

Is it possible to get a `previous_prediction` that is dense and `X` being sparse and thus the stacking will make something not expected?

Reply:
AFAICT it seems stacking a sparse with dense via `sp.hstack` gives you a sparse array (even though `sp.hstack` is not documented to support dense):

```bash
In [34]: from scipy.sparse import coo_matrix, hstack
    ...: 
    ...: A = coo_matrix([[1, 2], [3, 4]])

In [35]: B = np.zeros((2,2))

In [36]: hstack([A,B])
Out[36]: 
<2x4 sparse matrix of type '<class 'numpy.float64'>'
        with 4 stored elements in COOrdinate format>
```

Maybe from here: https://github.com/scipy/scipy/blob/f990b1d2471748c79bc4260baf8923db0a5248af/scipy/sparse/_construct.py#L654 ?

Reply:
Interesting. Probably this is the `np.hstack` that did not behave properly then.

Reply:
Yes you may be right. The [documentation](https://scipy.github.io/devdocs/reference/generated/scipy.sparse.hstack.html#scipy.sparse.hstack) does not mention that it would support sparse + dense.

I can't find any reference about whether this is intentional. This [scipy PR](https://github.com/scipy/scipy/pull/3057#issue-22512122) mentions that the stack functions converts everything to COO format (but we knew this from the code). Also, I found [this stackoverflow](https://stackoverflow.com/a/18417954) answer saying that you can stack sparse + dense.

Happy to add `previous_prediction` sparse conversion ? (Though this would be difficult to test since `sp.hstack` allows sparse + dense)

Reply:
Also a quick look at our code, I could not find any other cases where it would be possible to be stacking dense + sparse, if that helps the decision in whether we need to covert to sparse.